### ISSUE TITLE
V2 expansions.expand of "simple" geom3 crashes

### ISSUE THREAD

**Issue by kbruneel** [jscad/OpenJSCAD.org#876]
## Expected Behavior
For the expand not to crash

## Actual Behavior
It slowly fills up the memory until the browser crashes.

## Steps to Reproduce the Problem
I have trouble reproducing it and needs some help with this. Currently I play with some user inputs until the browser freezes. Then I click the pause button in the javascript console and use the call stack to find the expand in which it happens. It's always the same one.

I have tried to reproduce the behavior in an isolated test by copying the data from the geometry that is the input to the failing expand. This results in the following test code:

```
        var polies = [
            jscad.geometries.poly3.fromPointsAndPlane([[-19.6100000, -0.8000000, 11.8550000], [-19.6100000, -0.8000000, -11.8550000], [-19.6100000, -2.7500000, -11.8550000], [-19.6100000, -2.7500000, 11.8550000], ], [-1, 0, 0, 19.61]),
            jscad.geometries.poly3.fromPointsAndPlane([[-19.2723800, -0.8000000, 11.8550000], [-19.2723800, -2.7500000, 11.8550000], [-17.3200000, -2.7500000, -10.0000000], [-17.3200000, -0.8000000, -10.0000000], ], [0.9960335028595003, 3.0983531671633697e-16, 0.08897899292211668, -18.141090198747712]),
            jscad.geometries.poly3.fromPointsAndPlane([[-17.3200000, -0.8000000, -10.0000000], [-17.3200000, -2.7500000, -10.0000000], [-14.1071400, -2.7500000, -11.8550000], [-14.1071400, -0.8000000, -11.8550000], ], [0.5000110003630168, -1.3211722663644432e-16, 0.8660190526287371, -17.320381052574824]),
            jscad.geometries.poly3.fromPointsAndPlane([[-17.3200000, -0.8000000, -11.8550000], [-17.3200000, -0.8000000, -10.0000000], [-14.1071400, -0.8000000, -11.8550000], ], [0, 1, -1.0483128474924838e-16, -0.8000000000000003]),
            jscad.geometries.poly3.fromPointsAndPlane([[-17.3200000, -0.8000000, -11.8550000], [-19.6100000, -0.8000000, -11.8550000], [-19.6100000, -0.8000000, 11.8550000], [-19.2723800, -0.8000000, 11.8550000], [-17.3200000, -0.8000000, -10.0000000], ], [0, 1, -1.0483128474924838e-16, -0.8000000000000003]),
            jscad.geometries.poly3.fromPointsAndPlane([[-17.3200000, -2.7500000, -10.0000000], [-17.3200000, -2.7500000, -11.8550000], [-14.1071400, -2.7500000, -11.8550000], ], [0, -1, 1.3977504633233108e-16, 2.75]),
            jscad.geometries.poly3.fromPointsAndPlane([[-17.3200000, -2.7500000, -10.0000000], [-19.2723800, -2.7500000, 11.8550000], [-19.6100000, -2.7500000, 11.8550000], [-19.6100000, -2.7500000, -11.8550000], [-17.3200000, -2.7500000, -11.8550000], ], [0, -1, 1.3977504633233108e-16, 2.75]),
            jscad.geometries.poly3.fromPointsAndPlane([[-19.2723800, -0.8000000, 11.8550000], [-19.6100000, -0.8000000, 11.8550000], [-19.6100000, -2.7500000, 11.8550000], [-19.2723800, -2.7500000, 11.8550000], ], [0, 0, 0.9999999999999999, 11.854999999999999]),
            jscad.geometries.poly3.fromPointsAndPlane([[-19.6100000, -0.8000000, -11.8550000], [-14.1071400, -0.8000000, -11.8550000], [-14.1071400, -2.7500000, -11.8550000], [-19.6100000, -2.7500000, -11.8550000], ], [0, 0, -1, 11.855])
        ];

        var sub = jscad.geometries.geom3.create(polies);

        var shape = jscad.expansions.expand({delta: 1.3, corners: 'round'}, sub);
```

The problem now is that this code does not fail. It smoothly calculates the expand.

Is there some state of the geom3 object that I'm missing?

Can you recommend another technique to debug this problem?

## Specifications

  - Version: "@jscad/modeling": "^2.5.0"
  - Platform: OS X
  - Environment: Chrome Version 91.0.4472.114 (Official Build) (x86_64) and Node v12.18.2

**Comment by z3dev** [comment]
@kbruneel thanks much for the nice example.

First, some answers…
- there’s no additional state. The call to geom3.create() looks correct.
- there are several things to look at further…

The expand() function also supports the ‘segments’ option. Have you tried adjusting this to something smaller, like 16?

There might be issues with previous functions. Are you using any booleans functions? Any other expand or extrusion functions?

**Comment by kbruneel** [comment]
@z3dev 

Thanks for your quick answer!

For 'segments', I'm currently using the default option, which is 12. When I lower it to 8 the problem seems to disappear. This is ok for me, but there is still something wrong when 'segments' is set to 12.

I'm using a subtract before doing the expand. Is it possible that the subtract changes the result of the expand when I copy the full state of the inputs to the expand function?

Things I have noticed:
* If you look closely, the geometry in example has coplanar polygons. Maybe this has something to do with it?
* The problem happens in the union of line 33 of expandGeom3.js.
* the second argument of this union 'expanded' has 450 ish polygons in the browser when it crashes while in the test 'expanded' has 156 polygons. So the difference seems to be in expandShell function. Is there any caching or state there?

Many thanks,
Karel

**Comment by hrgdavor** [comment]
>The problem now is that this code does not fail. It smoothly calculates the expand.

@kbruneel can you give an example that fails as you described ?

**Comment by kbruneel** [comment]
@hrgdavor That's what I'm trying to do by reconstructing the geometrie and passing it to the expand function, as I showed above. But somehow the reconstruction is not correct, or there is some hidden state that I don't know about. Maybe some stuff gets cached and is reused by jscad?

**Comment by z3dev** [comment]
Thanks for the additional information.

so, the results change by changing the segments. Interesting.

the expand does a massive union as the final step, which is why the function is so SLOW. There may be a subtle bug when building up the ‘corners’.

I also noticed that the precision of the points is rounded, e.g. -19.6100000
I think that the example should be adjusted to use the actual values, full precision. We know that precision is often the cause of boolean issues.
Can you get those values?

one more thing… let’s just use the points when creating the polygons by using poly3.fromPoints()

**Comment by kbruneel** [comment]
@z3dev It was indeed the precision. I didn't know your toString function was changing the precision. So, now with full precision, I get the following code that fails:

```
        var polies = [
            jscad.geometries.poly3.fromPoints([[-19.61, -0.7999999999999986, 11.855], [-19.61, -0.8000000000000015, -11.855], [-19.61, -2.7500000000000018, -11.855], [-19.61, -2.7499999999999982, 11.855]]),
            jscad.geometries.poly3.fromPoints([[-17.32, -2.75, 10], [-17.32, -2.7500000000000013, -10], [-17.32, -0.8000000000000014, -10], [-17.32, -0.7999999999999987, 10]]),
            jscad.geometries.poly3.fromPoints([[-16.863040644206997, -0.8000000000000015, -10.28], [-16.863040644206997, -2.7500000000000018, -10.28], [-14.292644267871385, -2.7500000000000018, -11.855000000000016], [-14.292644267871383, -0.8000000000000015, -11.855000000000018]]),
            jscad.geometries.poly3.fromPoints([[-17.319999999999993, -0.8000000000000015, -9.999999999999996], [-17.319999999999993, -2.7500000000000018, -9.999999999999996], [-16.87560702649131, -2.7500000000000018, -10.272299999999998], [-16.866696319053347, -0.8000000000000015, -10.277759999999997]]),
            jscad.geometries.poly3.fromPoints([[-16.863040644207004, -2.7500000000000013, -10.280000000000001], [-16.863040644207004, -0.8000000000000014, -10.280000000000001], [-16.86669631905335, -0.8000000000000012, -10.27776], [-16.875607026491313, -2.75, -10.272300000000001]]),
            jscad.geometries.poly3.fromPoints([[-14.107140000000015, -0.7999999999999987, 11.85500000000003], [-14.107140000000015, -2.7499999999999982, 11.85500000000003], [-17.319999999999975, -2.7499999999999982, 9.999999999999956], [-17.319999999999975, -0.7999999999999987, 9.999999999999956]]),
            jscad.geometries.poly3.fromPoints([[-17.32, -0.7999999999999988, 9.999999999999993], [-17.32, -0.7999999999999986, 11.855], [-14.107139999999994, -0.7999999999999986, 11.855]]),
            jscad.geometries.poly3.fromPoints([[-17.32, -0.800000000000001, -11.855], [-17.32, -0.8000000000000008, -10.000000000000078], [-14.292644267871482, -0.800000000000001, -11.855]]),
            jscad.geometries.poly3.fromPoints([[-17.32, -0.800000000000001, -11.855], [-19.61, -0.800000000000001, -11.855], [-19.61, -0.7999999999999986, 11.855], [-17.32, -0.7999999999999986, 11.855]]),
            jscad.geometries.poly3.fromPoints([[-17.32, -2.7500000000000013, -10.000000000000076], [-17.32, -2.7500000000000018, -11.855], [-14.292644267871482, -2.7500000000000018, -11.855]]),
            jscad.geometries.poly3.fromPoints([[-17.32, -2.7499999999999982, 11.855], [-17.32, -2.7499999999999987, 9.999999999999996], [-14.107139999999996, -2.7499999999999982, 11.855]]),
            jscad.geometries.poly3.fromPoints([[-17.32, -2.7499999999999982, 11.855], [-19.61, -2.7499999999999982, 11.855], [-19.61, -2.7500000000000018, -11.855], [-17.32, -2.7500000000000018, -11.855]]),
            jscad.geometries.poly3.fromPoints([[-14.107139999999996, -0.7999999999999986, 11.855], [-19.61, -0.7999999999999986, 11.855], [-19.61, -2.7499999999999982, 11.855], [-14.107139999999994, -2.7499999999999982, 11.855]]),
            jscad.geometries.poly3.fromPoints([[-19.61, -0.8000000000000015, -11.855], [-14.292644267871486, -0.8000000000000015, -11.855], [-14.292644267871482, -2.7500000000000018, -11.855], [-19.61, -2.7500000000000018, -11.855]])
        ];

        var sub = jscad.geometries.geom3.create(polies);

        var shape = jscad.expansions.expand({delta: 1.3, corners: 'round', segments: 12}, sub);

```

**Comment by z3dev** [comment]
Excellent! Thanks for staying with us, and helping out.

I’m not able to work on this until September. Does this need to be fixed soon?

**Comment by hrgdavor** [comment]
I have attemped a fix at: 
https://github.com/hrgdavor/OpenJSCAD.org/tree/fix-expand

the branch does not pass tests yet but try it anyway.

**Comment by hrgdavor** [comment]
@kbruneel can you check the branch with possible fix and let us know if it works now, or provide further test samples 🦊

**Comment by mister-jscad** [comment]
@kbruneel Here's a little hack for you. Basically, you mesh is fine but some of those points are not aligned. Luckily, there's some new functionality to modify meshes.

`
  sub = generalize({snap: true, simplify: true}, sub)
`
There are two steps being called here. The first, 'snap' determines the precision of the mesh and snaps all the points to the precision, eliminating small floating point errors. The second, 'simplify' merges co-planar polygons into larger polygons, and eliminates unnecessary points.

Try it out and let us know if there are further issues.

**Comment by hrgdavor** [comment]
@mister-jscad is generalize open to users ?
it is present in the jcad build under jscad.modifiers but is not part of the docs: https://openjscad.xyz/docs/
it seems to fix leftover issues in #877

**Comment by z3dev** [comment]
> @mister-jscad is generalize open to users ?

Yes. But the doc strings are incorrect so no documentation yet.

**Comment by hrgdavor** [comment]
added #904

**Comment by platypii** [comment]
This appears to have been fixed by #898

It's possible there are other issues. But I tried using the example given by @kbruneel and could show that it crashed on every version before #898, and renders correctly after.
